src/V3Timing.cpp

```diff
@@ -0,0 +1,785 @@
+        // Get this sentree's trigger scheduler
+        auto* const trigSchedp = getCreateSchedulerp(nodep->sensesp());
+        auto* const flp = nodep->fileline();
+        // Replace self with a 'co_await trigSched.trigger()'
+        AstNodeStmt* awaitp = new AstCMethodHard{
+            flp, new AstVarRef{flp, trigSchedp, VAccess::READWRITE}, "trigger"};
+        awaitp->dtypeSetVoid();
+        awaitp = new AstCAwait{flp, awaitp};
+        awaitp->statement(true);
+        if (nodep->stmtsp()) awaitp->addNext(nodep->stmtsp()->unlinkFrBackWithNext());
+        nodep->replaceWith(commentp);
+        commentp->addNextHere(awaitp);
+        nodep->deleteTree();
+    }
+    virtual void visit(AstDelay* nodep) override {
+        iterateChildren(nodep);
+        auto* const flp = nodep->fileline();
+        // Replace self with a 'co_await dlySched.delay(<lhsp>)'
+        AstNodeStmt* awaitp
+            = new AstCMethodHard{flp, new AstVarRef{flp, m_dlySchedp, VAccess::READWRITE}, "delay",
+                                 nodep->lhsp()->unlinkFrBack()};
+        awaitp->dtypeSetVoid();
+        awaitp = new AstCAwait{flp, awaitp};
+        awaitp->statement(true);
+        if (nodep->stmtsp()) awaitp->addNext(nodep->stmtsp()->unlinkFrBackWithNext());
+        nodep->replaceWith(awaitp);
+        nodep->deleteTree();
+    }
+    virtual void visit(AstCAwait* nodep) override {
+        if (m_cfuncp) m_cfuncp->user1(true);
+    }
+    virtual void visit(AstCFunc* nodep) override {
+        m_cfuncp = nodep;
+        iterateChildren(nodep);
+        if (nodep->isCoroutine()
+            && !nodep->user1()) {  // If coroutine but no awaits (e.g. a class method that
+                                   // overrides/is overriden by a suspendable, but doesn't have any
+                                   // awaits itself)
+            // Add a co_return at the end so the compiler doesn't complain
+            nodep->addStmtsp(new AstCStmt{nodep->fileline(), "co_return;\n"});
+            nodep->user1(true);  // So we won't add another
+        }
+        m_cfuncp = nullptr;
+        m_lastTimingSensp = nullptr;
+    }
+    virtual void visit(AstNodeProcedure* nodep) override {
+        iterateChildren(nodep);
+        m_lastTimingSensp = nullptr;
+    }
+    virtual void visit(AstNodeVarRef* nodep) override {
+        if (m_lastTimingSensp && nodep->access().isWriteOrRW()) {
+            auto*& senTreep = m_timingKit.varToSenTreep[nodep->varScopep()];
+            if (senTreep) {
+                if (senTreep->user1()) {
+                    // Clone it so we don't modify a global sentree
+                    senTreep = senTreep->cloneTree(false);
+                    senTreep->user1p(nullptr);
+                }
+                senTreep->addSensesp(m_lastTimingSensp->sensesp()->cloneTree(true));
+            } else {
+                senTreep = m_lastTimingSensp;
+            }
+        }
```

Comment:
This is on the right track, but there are a lot of exciting bugs lurking here. consider:

```systemverilog
always @(posedge clk) begin
  if (foo)
    @(a);
  else
    @(b);
  x = y; // <- might be modified under both @(a) and @(b), but your current implementation only picks up @(b).
end
```

Also I think we need to trace function calls:
```systemverilog
always @(posedge clk) begin
  @(a);
  f(); // < What might this write to?
end

For correctness we need to be either precise or conservative.

Reply:
Now we do all timing controls in the process.